tests/fixtures/bidaf/experiment.json

```diff
@@ -46,7 +46,7 @@
       "num_layers": 1
     },
     "similarity_function": {
```

Comment:
And in these places we'd want to change this from a "similarity_function" key to a "matrix_attention" key.  This means also changing the `BiDAF` model to take a `MatrixAttention` object instead of a `SimilarityFunction` object in its constructor.

Reply:
Again as stated in my comment above this would break backwards compatibility. Is that ok? 

In this current PR the bidaf does take a MatrixAttention object instead of a SimilarityFunction. 
In the from_params method it converts these similarity_function into the correct MatrixAttention impl without compromising backwards compatibility.

Reply:
Yes, this will break backwards compatibility of the json configuration files.  That's ok.

Reply:
Actually, another idea is just to keep the `BiDAF` and other model code as it is, but use `LegacyMatrixAttention()` to construct the attention instead of `MatrixAttention`, in the constructor.  Then no fixtures need to be updated, the sniff tests still pass, and everything is ok.  New code can still use the nicer API, but we don't break anything at all.

Reply:
If I understand you correctly you are saying
1. the constructor to the BiDAF would still take in a similarity function (so there is no breaking change).
2. Within the constructor it will create LegacyMatrixAttention. 


If we do this then the bidaf will be bound to the slow performance of the legacy impl.
Is that ok?

Reply:
For this PR, yes.  That way we have a minimal change here that introduces a new API.  Then in a future PR, we can think about updating the code (and our sniff tests, and everything) to use the new API.

Reply:
Cool